src/Database/Franz/Reader.hs

```diff
@@ -240,3 +268,16 @@ handleQuery FranzReader{..} dir (Query name begin_ end_ rt) = do
           let body' = maybe id (IM.insert val) lastItem body
           return $! maybe minBound fst $ IM.maxView body'
     return $! range begin end rt allOffsets
+  where
+    acquire = do
+      allStreams <- readTVarIO vStreams
+      let !path = prefix </> dir </> B.unpack name
+      let !streams = maybe mempty id $ HM.lookup dir allStreams
+      stream <- case HM.lookup name streams of
+        Nothing -> do
+          s <- createStream watchManager path
+          atomically $ modifyTVar' vStreams $ HM.insert dir $ HM.insert name s streams
+          return s
+        Just vStream -> return vStream
+      addActivity stream
```

Comment:
You want to take a lock on vStreams and `addActivity` inside, while still holding a lock. Otherwise we may get `stream` that is killed by reaper before we `addActivity` here. If you modify the activity inside the transaction, either:

* reaper is first, it deletes the stream, `lookup` here fails and we re-create the stream
* reaper is second, it blocks until transaction here finishes, after transaction the activity is updated, stream is left intact.

Reply:
Fixed